cmd/http-tracer.go

```diff
@@ -174,3 +176,63 @@ func Trace(f http.HandlerFunc, logBody bool, w http.ResponseWriter, r *http.Requ
 	t.RespInfo = rs
 	return t
 }
+
+// Trace writes http trace to writer
+func httpTrace(doneCh chan struct{}, trcAll bool) chan []byte {
+	traceCh := make(chan []byte)
+	go func() {
+		defer close(traceCh)
+
+		remoteHosts := getRemoteHosts(globalEndpoints)
+		peers, err := getRestClients(remoteHosts)
+		if err != nil {
+			return
+		}
+
+		var wg = &sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			buf := &bytes.Buffer{}
+			ch := globalHTTPTrace.Subscribe()
+			defer globalHTTPTrace.Unsubscribe(ch)
+			for {
+				select {
+				case entry := <-ch:
+					trcInfo := entry.(trace.Info)
+					path := strings.TrimPrefix(trcInfo.ReqInfo.Path, "/")
+					// omit inter-node traffic if trcAll is false
+					if !trcAll && strings.HasPrefix(path, minioReservedBucket) {
+						continue
+					}
+					buf.Reset()
+					enc := json.NewEncoder(buf)
+					enc.SetEscapeHTML(false)
+					if err := enc.Encode(trcInfo); err != nil {
+						continue
+					}
+					traceCh <- buf.Bytes()
+				case <-doneCh:
```

Comment:
There appears to be a deadlock corner case here, for e.g. consider the situation:

1. TraceHandler's select block has an error and the handler calls close(doneCh) (due to the defer statement).
2. At that time, this select block's first case statement may be blocked on line 215 trying to send on traceCh (there is no receiver on the other end, so it blocks).
3. So the second case statement (line 216) never fires and this thread never exits.

Similar variation can happen in the go routines launched in lines 224-234.

Reply:
yes, good catch. There is similar race condition in pubsub.go where we can endup trying to send to a closed channel

Reply:
can you check now?